### PR TITLE
Adds kapt apoptions to kotlin_library

### PR DESCRIPTION
**Pull request by mmdango**
This will allow users to pass in kapt annotation processor options to kapt.

Not sure if the way I wrote the docs is okay, but I don't think there are any examples of Hash type params so I wanted to be explicit.

Slack discussion here: https://buckbuild.slack.com/archives/C6N81MF55/p1537224378000100

### CODE REVIEW

**Comment by mmdango** [comment]
cc @thalescm @cwoodwar6

File: docs/rule/kotlin_library.soy [hunk 1]
```diff
@@ -32,7 +32,7 @@ of the compiled class files and resources.
 
 {call buck.arg}
   {param name: 'srcs' /}
-  {param default: '[]' /}
+  {param default: '{}' /}
```
Comment (cwoodwar6) [inline docs/rule/kotlin_library.soy]: Was the intentional? Seems that you want this as part of your `kapt_apoptions`

Comment (mmdango) [reply docs/rule/kotlin_library.soy]: oof, good catch. yeah, i meant it for the kapt_apoptions param. thanks!

File: docs/rule/android_library.soy [hunk 2]
```diff
@@ -171,6 +171,20 @@ compiled class files and resources.
   {/param}
 {/call}
 
+{call buck.arg}
+  {param name: 'kapt_apoptions' /}
```
Comment (kageiit) [inline docs/rule/android_library.soy]: kapt_apoptions -> kapt_ap_options

**Comment by sbalabanov** [comment]
tests are failing, seems like soy template is incorrect

**Comment by mmdango** [comment]
@sbalabanov I believe I fixed the soy tests, but looks like there are unrelated tests that are failing in AppVeyor. I'm also a bit confused what the error message for the Travis CI is: https://travis-ci.org/facebook/buck/jobs/439005432#L7325 it just seems like the script always throws 1...

**Comment by thalescm** [comment]
Another point is. If we open an API just for ap_options. it seems fair to open another one for every new type of thing the compiler accepts (and the old ones too) and not having the `extra_kotlinc_argmunts` at all. Which can lead us to a very confusing and over granular API that is hard to use and hard to maintain.

File: src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java [hunk 3]
```diff
@@ -65,6 +65,7 @@ public CompileToJarStepFactory configure(
         kotlinBuckConfig.getKotlinHomeLibraries(),
         kotlinArgs.getExtraKotlincArguments(),
         kotlinArgs.getAnnotationProcessingTool().orElse(AnnotationProcessingTool.KAPT),
+        kotlinArgs.getKaptApOptions(),
```
Comment (raviagarwal7) [inline src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java]: would be good to support macros as well in https://github.com/facebook/buck/commit/3e0b4cc444d86f565124807f729bbe7d9748c83c#diff-8bd16d403f308d9df417653053234adb in kaptApOptions

**Comment by thalescm** [comment]
@mmdango would you mind rebasing it?

**Comment by thalescm** [comment]
@mmdango Seems like there are still formatting issues.
What I normally do (to fix them all automatically) is to call [this script](https://github.com/facebook/buck/blob/master/scripts/travisci_test_java_file_forma)
But first deleting lines:
4-10;
32-33;

Also: Don't forget to replace `${TRAVIS_BRANCH}` on line 21 to whatever sha is the master you're pointing to (e.g `$(git rev-parse master)`.

At the end you'll have all necessary changes, just add them and commit :)

**Comment by mmdango** [comment]
@thalescm done

**Comment by thalescm** [comment]
@sbalabanov Can we get this merged please?

**Comment by thalescm** [comment]
@styurin You imported this more than two weeks ago. Do you have any updates?

**Comment by styurin** [comment]
Could you rebase it? There are some merge conflicts.

**Comment by mmdango** [comment]
@styurin rebased. sorry for the delay, just got back from vacation

**Pull request closed without merging**